File: java/src/main/java/geospat/GeoMatrix.java
package geospat;

import java.util.HashMap;
import java.util.ArrayList;

public class GeoMatrix<T> {
    //-----------------------------------------------------------------------------------
    public enum GeoMatrixAccuracy {
        HIGH,
        MEDIUM,
        LOW
    }
    //-----------------------------------------------------------------------------------
    private class GeoMatrixItemInfo {
        private Polygon polygon_;
        private T data_;
        GeoMatrixItemInfo(final Polygon p,final T data){
            data_=data;
            polygon_=p;
        }
    }
    //-----------------------------------------------------------------------------------
    private class GeoMatrixItemList {
        public ArrayList<GeoMatrixItemInfo> infoList_=new ArrayList<GeoMatrixItemInfo>();
    }
    //-----------------------------------------------------------------------------------
    private ArrayList<GeoMatrixItemInfo> items_= new ArrayList<GeoMatrixItemInfo>();
    private Rectangle bBox_;
    private final int row_=1000;
    private HashMap<Point,GeoMatrixItemList>  map_=new HashMap<Point,GeoMatrixItemList>((row_*row_));

    //-----------------------------------------------------------------------------------
    private Point getMatrixPoint(final Point in) {
        Point ext = bBox_.getLength();
        double slotX = (ext.getX() / row_);
        double slotY = (ext.getY() / row_);

        double x = ((in.getX() - bBox_.getStart().getX()) / slotX);
        double y = ((in.getY() - bBox_.getStart().getY()) / slotY);

        x =  Math.min(x, row_-1);
        y =  Math.min(y, row_-1);


        Point out=new Point((int)x,(int)y);
        return out;
    }
    
    //-----------------------------------------------------------------------------------
    private void coverArea(final Point start, final Point end, GeoMatrixItemInfo info) {
        for (int x = (int)start.getX(); x <= end.getX(); x++) {
            for (int y = (int)start.getY(); y <= end.getY(); y++) {
                Point cPoint=new Point(x, y);         
                
                GeoMatrixItemList list=map_.get(cPoint);
                if(list==null){
                    list=new GeoMatrixItemList();
                    map_.put(cPoint,list);
                }
                list.infoList_.add(info);
            }
        }
    }
    
    //-----------------------------------------------------------------------------------
    public void add(final Polygon poly, final T data) {
        Rectangle rect = poly.getBoundingBox();

        if (items_.isEmpty()) {
            bBox_ = new Rectangle(rect);
        } else {
            bBox_.add(rect);
        }
         items_.add(new GeoMatrixItemInfo(poly,data));
    }
    
    //-----------------------------------------------------------------------------------
    public void build(){
        for (GeoMatrixItemInfo i : items_) {
            Rectangle rect = i.polygon_.getBoundingBox();
            Point start=getMatrixPoint(rect.getStart());
            Point end=getMatrixPoint(rect.getEnd());
            coverArea(start, end, i);
        }
        System.out.printf("load %d  items\n",map_.size());
    }
    
    //-----------------------------------------------------------------------------------
    public boolean query(final Point point,ArrayList<T> out,GeoMatrixAccuracy acc,int maxItems){
        Point cPoint=getMatrixPoint(point);
        
        GeoMatrixItemList list=map_.get(cPoint);
        if(list==null){
            return false;
        }

        if(list.infoList_.size()==1){
            out.add(list.infoList_.get(0).data_);
            return true;
        }
        
        for(GeoMatrixItemInfo info : list.infoList_){

            if(out.size()==maxItems){
                break;
            }

            switch(acc){
                case LOW:
                    out.add(info.data_);
                    break;

                case MEDIUM:
                    if(info.polygon_.getBoundingBox().pointInside(point)){
                        out.add(info.data_);
                    }
                    break;

                case HIGH:
                    if(info.polygon_.pointIsInside(point)){
                        out.add(info.data_);
                        return true;
                    }
                    break;
            }
        }
        return true;
    }

    //-----------------------------------------------------------------------------------
    public ArrayList<T> query(final Point point,GeoMatrixAccuracy acc,int maxItems){
        ArrayList<T> out=new ArrayList<T>();
        query(point,out,acc,maxItems);
        return out;
    }
    //-----------------------------------------------------------------------------------
}

File: rust/polygon_search/src/geo/geo_matrix.rs
use crate::geo::point::Point;
use crate::geo::polygon::Polygon;
use crate::geo::rectangle::Rect;
use std::collections::hash_map::HashMap;

pub enum GeoMatrixAccuracy {
    High,
    Medium,
    Low,
}

type MatrixPoint = (u64, u64);

const ROW: f64 = 1000.0;

pub struct GeoMatrix<T>
where
    T: Clone,
{
    map_: HashMap<MatrixPoint, Vec<usize>>,
    objects_list_: Vec<(Polygon, T)>,
    bbox_: Rect,
}

impl<T> GeoMatrix<T>
where
    T: Clone,
{
    pub fn new() -> GeoMatrix<T> {
        return GeoMatrix {
            map_: HashMap::new(),
            objects_list_: Vec::new(),
            bbox_: Rect::new(Point::new(0.0, 0.0), Point::new(0.0, 0.0)),
        };
    }

    fn get_matrix_point(bbox: &Rect, p: &Point) -> MatrixPoint {
        let ext = bbox.get_length();
        let slotx = ext.get_x() / ROW;
        let sloty = ext.get_y() / ROW;

        let mut x = (p.get_x() - bbox.get_start().get_x()) / slotx;
        let mut y = (p.get_y() - bbox.get_start().get_y()) / sloty;

        x = x.min(ROW);
        y = y.min(ROW);

        return (x as u64, y as u64);
    }

    fn cover_area(&mut self, start: &MatrixPoint, end: &MatrixPoint, index: usize) {
        for x in start.0..(end.0+1) {
            for y in start.1..(end.1+1) {
                let point = (x, y);
                match self.map_.get_mut(&point) {
                    Some(v) => {
                        v.push(index);
                    }
                    None => {
                        let v = vec![index];
                        self.map_.insert(point, v);
                    }
                }
            }
        }
    }

    pub fn add(&mut self, pol: &Polygon, data: &T) {
        let mut temp_pol = pol.clone();
        let bbox = temp_pol.get_bounding_box();
        if self.objects_list_.len() == 0 {
            self.bbox_ = *bbox;
        } else {
            self.bbox_.add(bbox);
        }
        self.objects_list_.push((pol.clone(), data.clone()));
    }

    pub fn build(&mut self) {
        let l = self.objects_list_.len();
        for i in 0..l {
            let (pol, _) = &mut self.objects_list_[i];
            let rect = pol.get_bounding_box();
            let start = GeoMatrix::<T>::get_matrix_point(&self.bbox_, rect.get_start());
            let end = GeoMatrix::<T>::get_matrix_point(&self.bbox_, rect.get_end());
            self.cover_area(&start, &end, i);
        }
    }

    pub fn query(
        &mut self,
        point: &Point,
        out: &mut Vec<T>,
        acc: &GeoMatrixAccuracy,
        max_count: usize,
    ) -> bool {
        let mp = GeoMatrix::<T>::get_matrix_point(&self.bbox_, point);

        let copy_value = |index: &usize, out: &mut Vec<T>| {
            out.push(self.objects_list_[*index].1.clone());
        };

        let res = self.map_.get(&mp);
        if res == None {
            return false;
        }
        let items = res.unwrap();

        if items.len() == 1 {
            copy_value(&items[0], out);
            return true;
        }

        for i in items {
            let (pol, _) = &self.objects_list_[*i];

            match acc {
                GeoMatrixAccuracy::High => {
                    if pol.point_is_inside(point) {
                        copy_value(i, out);
                        return true;
                    }
                }
                GeoMatrixAccuracy::Medium => {
                    if pol.point_is_in_bbox(point) {
                        copy_value(i, out);
                    }
                }
                GeoMatrixAccuracy::Low => {
                    copy_value(i, out);
                }
            }

            if out.len() == max_count {
                return true;
            }
        }
        return true;
    }
}
